src/main/java/io/debezium/connector/yugabytedb/consistent/Merger.java

```diff
@@ -19,17 +24,36 @@ public class Merger {
     public Merger(List<String> tabletList) {
         tabletList.forEach(tabletId -> {
             mergeSlots.put(tabletId, new ArrayList<>());
-            tabletSafeTime.put(tabletId, BigInteger.ZERO);
+            setTabletSafeTime(tabletId, BigInteger.ZERO);
         });
     }
 
+    /**
+     * Add the passed message to the priority queue as well as the respective merge slot for
+     * further processing. This also updates the tablet safetime after each message is added to the
+     * queue.
+     * @param message the message to be added to the queue
+     * @see Message
+     */
     public synchronized void addMessage(Message message) {
+        setTabletSafeTime(message.tablet, message.commitTime);
         if (message.record.getRowMessage().getOp() == CdcService.RowMessage.Op.SAFEPOINT) {
             LOGGER.debug("Received safe point message {}", message);
-            tabletSafeTime.put(message.tablet, message.commitTime);
-            LOGGER.debug("Put {}:{}, verifying {}", message.tablet, message.commitTime, tabletSafeTime.get(message.tablet));
             return;
         }
+
+        // TODO: Wrap these checks under a flag maybe.
+        if (!isMergeSlotSorted(this.mergeSlots.get(message.tablet))) {
+            throw new AssertionError("Merge slot is not sorted");
+        }
+
+        if (!this.mergeSlots.get(message.tablet).isEmpty()
```

Comment:
It is better to compare the two messages and not just the commitTime. I think this condition is not correct as it may break for BEGIN/COMMIT etc. `message.compareTo(lastMessage)` is more appropriate.